src/io/parquet/read/mod.rs

```diff
@@ -314,44 +314,33 @@ fn page_iter_to_array<I: FallibleStreamingIterator<Item = DataPage, Error = Parq
-                            iter,
-                            DataType::FixedSizeBinary(*n as usize),
-                            metadata,
-                        )
-                        .map(|e| {
-                            let a = e
-                                .into_iter()
-                                .map(|v| {
-                                    v.and_then(|v1| {
-                                        // Pad with the value of the MSB to correctly handle (two's complement) negative integers.
-                                        let msb_set = v1.first().unwrap_or(&0) >> 7 == 1;
-                                        let padding = if msb_set {
-                                            &ones_padding
-                                        } else {
-                                            &zeros_padding
-                                        };
-                                        [padding, v1]
-                                            .concat()
-                                            .try_into()
-                                            .map(i128::from_be_bytes)
-                                            .ok()
-                                    })
-                                })
-                                .collect::<Vec<_>>();
-                            Box::new(PrimitiveArray::<i128>::from(a).to(data_type))
-                                as Box<dyn Array>
-                        })
-                    }
+                &PhysicalType::FixedLenByteArray(n) if n > 16 => {
+                    Err(ArrowError::NotYetImplemented(format!(
+                        "Can't decode Decimal128 type from Fixed Size Byte Array of len {:?}",
+                        n
+                    )))
+                }
+                &PhysicalType::FixedLenByteArray(n) => {
+                    let n = n as usize;
+                    let fixed_size_binary_array = fixed_size_binary::iter_to_array(
+                        iter,
+                        DataType::FixedSizeBinary(n),
+                        metadata,
+                    )?;
+                    let i128_values =
+                        fixed_size_binary_array
+                            .into_iter()
+                            .map(|value: Option<&[u8]>| {
+                                // Copy the fixed-size byte value to the start of a 16 byte stack
+                                // allocated buffer, then use an arithmetic right shift to fill in
+                                // MSBs, which accounts for leading 1's in negative (two's complement)
+                                // values.
+                                let mut bytes = [0u8; 16];
+                                bytes[..n].copy_from_slice(value?);
+                                Some(i128::from_be_bytes(bytes) >> 8 * (16 - n))
+                            });
+                    let i128_array = PrimitiveArray::<i128>::from_iter(i128_values);
```

Comment:
```suggestion
                    let values =
                        fixed_size_binary_array
                            .values()
                            .iter()
                            .map(|value: &[u8]| {
                                // Copy the fixed-size byte value to the start of a 16 byte stack
                                // allocated buffer, then use an arithmetic right shift to fill in
                                // MSBs, which accounts for leading 1's in negative (two's complement)
                                // values.
                                let mut bytes = [0u8; 16];
                                bytes[..n].copy_from_slice(value);
                                i128::from_be_bytes(bytes) >> 8 * (16 - n)
                            })
                            .collect::<Vec<_>>();
                    let validity = fixed_size_binary_array.validity().cloned();
                    let i128_array = PrimitiveArray::<i128>::from_data(i128::DATA_TYPE, values.into(), validity);
```

since this operator is infallible, we can bypasse (the expensive) iteration and compute of a new validity.